File: ks-codes/ksintegrate.cpp
typedef std::complex<double> Complex;

// ksintegrate: integrate kuramoto-sivashinsky equation (Python)
//       u_t = -u*u_x - u_xx - u_xxxx, domain x in [0,Lx], periodic BCs 

//  inputs
//        u = initial condition (vector of u(x) values on uniform gridpoints))
//       Lx = domain length
//       dt = time step
//       Nt = number of integration timesteps
//    nsave = save every nsave-th time step
// 
//  outputs
//        u = final state, vector of u(x, Nt*dt) at uniform x gridpoints
   
void ksintegrate(Complex* u, const int Nx, const double Lx, const double dt, const int Nt) {

  int* kx        = new int[Nx];     // for Nx=8, kx = [0 1 2 3 0 -3 -2 -1]
  double*  alpha = new double[Nx];  // real wavenumbers:    exp(alpha*x)
  Complex* D     = new Complex[Nx]; // D = d/dx operator in Fourier space
  Complex* G     = new Complex[Nx]; // -1/2 D operator in Fourier space
  double*  L     = new double[Nx];    // -D^2 - D^4

  // Assign Fourier wave numbers in order that FFTW produces
  for (int n=0; n<Nx/2; ++n)      
    kx[n] = n;

  kx[Nx/2] = 0.0;                 

  for (int n=Nx/2+1; n<Nx; ++n)   
    kx[n] = -Nx + n;


  // Assign alpha, D, G, L
  const double a = 2*M_PI/Lx;
  for (int n=0; n<Nx; ++n) {   
    alpha[n] = a*kx[n];   
    double alpha2 = alpha[n]*alpha[n];  // alpha[n]^2
    L[n] = alpha2*(1-alpha2);
    G[n] = Complex(0.0, -0.5*alpha[n]);
  }

  // CNAB formula
  // (I - dt/2 L) u^{n+1} = (I + dt/2 L) u^n + 3dt/2 N^n - dt/2 N^{n-1}
  //
  // u^{n+1} = (I - dt/2 L)^{-1} [(I + dt/2 L) u^n + 3dt/2 N^n - dt/2 N^{n-1}]
  //
  // let A = (I + dt/2 L)
  // let B = (I - dt/2 L)^{-1}

  // some convenience variables
  double dt2  = dt/2;
  double dt32 = 3*dt/2;
  double Nx_inv = 1.0/Nx;
  double Nx_inv2 = Nx_inv*Nx_inv;

  double* A = new double[Nx];
  double* B = new double[Nx];

  for (int n=0; n<Nx; ++n) {
    A[n] = 1.0 + dt2*L[n];
    B[n] = 1.0/(1.0 - dt2*L[n]);
  }

  Complex* uu = new Complex[Nx];  
  for (int n=0; n<Nx; ++n) 
    uu[n] = u[n]*u[n];

  // Cast Native C++ complex array pointers to FFTW equivalent.
  fftw_complex* u_fftw  =  reinterpret_cast<fftw_complex*>(u);
  fftw_complex* uu_fftw =  reinterpret_cast<fftw_complex*>(uu);
  
  // Construct FFTW plans
  fftw_plan u_fftw_plan   = fftw_plan_dft_1d(Nx, u_fftw,  u_fftw,  FFTW_FORWARD, FFTW_ESTIMATE);
  fftw_plan u_ifftw_plan  = fftw_plan_dft_1d(Nx, u_fftw,  u_fftw,  FFTW_BACKWARD, FFTW_ESTIMATE);
  fftw_plan uu_fftw_plan  = fftw_plan_dft_1d(Nx, uu_fftw, uu_fftw, FFTW_FORWARD,  FFTW_ESTIMATE);
  fftw_plan uu_ifftw_plan = fftw_plan_dft_1d(Nx, uu_fftw, uu_fftw, FFTW_BACKWARD, FFTW_ESTIMATE);

  // Compute  Nn = G*fft(u*u), (-u u_x, spectral)
  fftw_execute(uu_fftw_plan);   // uu physical -> spectral

  Complex* Nn = new Complex[Nx];  
  for (int n=0; n<Nx; ++n)      // compute N(u) = -1/2 d/dx u^2, spectral
    Nn[n] = G[n]*uu[n];

  Complex* Nn1 = new Complex[Nx];  

  fftw_execute(u_fftw_plan);    // transform u physical -> spectral

  // timestepping loop
  for (int s=0; s<Nt; ++s) {

    for (int n=0; n<Nx; ++n) {
      Nn1[n] = Nn[n];           // shift nonlinear term in time: N^{n-1} <- N^n
      uu[n] = u[n];             // copy u to uu (spectral)
    }      

    fftw_execute(uu_ifftw_plan); // uu (holding u) spectral -> physical
    for (int n=0; n<Nx; ++n) 
      uu[n] *= Nx_inv;           // normalize

    for (int n=0; n<Nx; ++n) 
      uu[n] = uu[n]*uu[n];       // compute uu = u^2 physical,

    fftw_execute(uu_fftw_plan);  // transform uu physical -> spectral

    for (int n=0; n<Nx; ++n) 
      Nn[n] = G[n]*uu[n];        // compute Nn == -u u_x, spectral
    
    for (int n=0; n<Nx; ++n) 
      u[n] = B[n] * (A[n] * u[n] + dt32*Nn[n] - dt2*Nn1[n]);

  }

  fftw_execute(u_ifftw_plan);    // transform uu physical -> spectral
  for (int n=0; n<Nx; ++n) 
    u[n] *= Nx_inv;              // normalize  

  fftw_destroy_plan(u_fftw_plan);
  fftw_destroy_plan(uu_fftw_plan);
  fftw_destroy_plan(u_ifftw_plan);
  fftw_destroy_plan(uu_ifftw_plan);

  delete[] kx;
  delete[] alpha;
  delete[] G;
  delete[] L;
  delete[] A;
  delete[] B;
  delete[] uu;
  delete[] Nn;
  delete[] Nn1;
}

File: ks-codes/ksintegrate.jl
function ksintegrateInplace(u, Lx, dt, Nt, nsave);
    u = (1+0im)*u                       # force u to be complex
    Nx = length(u)                      # number of gridpoints
    kx = vcat(0:Nx/2-1, 0:0, -Nx/2+1:-1)# integer wavenumbers: exp(2*pi*kx*x/L)
    alpha = 2*pi*kx/Lx                  # real wavenumbers:    exp(alpha*x)
    D = 1im*alpha                       # spectral D = d/dx operator 
    L = alpha.^2 - alpha.^4             # spectral L = -D^2 - D^4 operator
    G = -0.5*D                          # spectral -1/2 D operator

    Nsave = div(Nt, nsave)+1            # number of saved time steps
    t = (0:Nsave)*(dt*nsave)            # t timesteps
    U = zeros(Nsave, Nx)                # matrix of u(xⱼ, tᵢ) values
    U[1,:] = u                          # assign initial condition to U
    s = 2                               # counter for saved data
 
    # some convenience variables
    dt2  = dt/2
    dt32 = 3*dt/2
    A_inv = (ones(Nx) - dt2*L).^(-1)
    B     =  ones(Nx) + dt2*L
    
    # compute in-place FFTW plans
    FFT! = plan_fft!(u, flags=FFTW.ESTIMATE)
    IFFT! = plan_ifft!(u, flags=FFTW.ESTIMATE)

    # compute nonlinear term Nn = -u u_x 
    Nn  = G.*fft(u.^2);    # Nn = -1/2 d/dx u^2 = -u u_x
    Nn1 = copy(Nn);        # Nn1 = Nn at first time step
    FFT!*u;                # transform u physical -> spectral
    
    # timestepping loop
    for n = 1:Nt

        Nn1 .= Nn       # shift nonlinear term in time

        Nn .= u         # put u into Nn 
        IFFT!*Nn;       # transform Nn = u to gridpt values, in place
        Nn .= Nn.*Nn;   # collocation calculation, set Nn = u^2
        FFT!*Nn;        # transform Nn = u^2 back to spectral coeffs
        Nn .= G.*Nn;    # apply G = -1/2d/dx  to compute Nn = -1/2 d/dx (u^2)

        # loop fusion! Julia translates this line into a single for-loop on
        #   u[i] = A_inv[i] * (B[i]*u[i] + dt32*Nn[i] - dt2*Nn1[i]; 
        # no temporary vectors! 

        u .= A_inv .* (B .* u .+ dt32.*Nn .- dt2.*Nn1); 
        
        if mod(n, nsave) == 0
            U[s,:] = real(ifft(u))
            s += 1            
        end
    end
   
    t,U
end
